### PR TITLE
Use random apiserver address when dialing

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

We want to ensure one controller in an HA setup is not swamped with connections.
So we shuffle the api addresses prior to dialling.

## QA steps

Bootstrap, enable HA, and run some Juju operations.

### CODE REVIEW

File: api/apiclient.go [hunk 1]
```diff
@@ -591,14 +592,27 @@ func dialAPI(info *Info, opts0 DialOpts) (*dialResult, error) {
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	dialInfo, err := dialWebsocketMulti(ctx, info.Addrs, path, opts)
+	dialInfo, err := dialWebsocketMulti(ctx, shuffleAddresses(info.Addrs), path, opts)
```
Comment (axw) [inline api/apiclient.go]: I'm -1 on applying this logic to client (non-agent) connections. Perhaps move this logic to worker/apicaller?

File: api/apiclient.go [hunk 2]
```diff
@@ -591,14 +592,27 @@ func dialAPI(info *Info, opts0 DialOpts) (*dialResult, error) {
...
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	logger.Infof("connection established to %q", dialInfo.urlStr)
 	return dialInfo, nil
 }
 
+func init() {
+	rand.Seed(time.Now().UnixNano())
```
Comment (axw) [inline api/apiclient.go]: assuming you do move the code to worker/apicaller, please create a worker-specific rand.Rand rather than modifying global state

Comment (wallyworld) [inline api/apiclient.go]: done

File: api/apiclient.go (hunk 1 above)
Comment (wallyworld) [inline api/apiclient.go]: moved to apicaller

File: worker/apicaller/connect.go [hunk 3]
```diff
@@ -83,12 +93,18 @@ func connectFallback(
 	conn api.Connection, didFallback bool, err error,
 ) {
 
+	infoCopy := *info
+	if len(infoCopy.Addrs) > 1 {
+		src := rand.NewSource(time.Now().UnixNano())
```
Comment (axw) [inline worker/apicaller/connect.go]: the Rand should really live longer than a single connection call, but I guess it's not really a big deal

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**